kmath-tensors/src/commonMain/kotlin/space/kscience/kmath/tensors/core/utils.kt

```diff
@@ -0,0 +1,104 @@
+package space.kscience.kmath.tensors.core
+
+import space.kscience.kmath.nd.as1D
+import space.kscience.kmath.samplers.GaussianSampler
+import space.kscience.kmath.stat.RandomGenerator
+import space.kscience.kmath.structures.*
+import kotlin.math.*
+
+/**
+ * Returns a reference to [IntArray] containing all of the elements of this [Buffer] or copy the data.
+ */
+internal fun Buffer<Int>.array(): IntArray = when (this) {
+    is IntBuffer -> array
+    else -> this.toIntArray()
+}
+
+
+/**
+ * Returns a reference to [DoubleArray] containing all of the elements of this [Buffer] or copy the data.
+ */
+internal fun Buffer<Double>.array(): DoubleArray = when (this) {
+    is DoubleBuffer -> array
+    else -> this.toDoubleArray()
+}
+
+internal inline fun getRandomNormals(n: Int, seed: Long): DoubleArray {
+    val distribution = GaussianSampler(0.0, 1.0)
+    val generator = RandomGenerator.default(seed)
+    return distribution.sample(generator).nextBufferBlocking(n).toDoubleArray()
+}
+
+internal inline fun getRandomUnitVector(n: Int, seed: Long): DoubleArray {
+    val unnorm = getRandomNormals(n, seed)
+    val norm = sqrt(unnorm.map { it * it }.sum())
+    return unnorm.map { it / norm }.toDoubleArray()
+}
+
+internal inline fun minusIndexFrom(n: Int, i: Int) : Int = if (i >= 0) i else {
+    val ii = n + i
+    check(ii >= 0) {
+        "Out of bound index $i for tensor of dim $n"
+    }
+    ii
+}
+
+internal inline fun <T> BufferedTensor<T>.minusIndex(i: Int): Int =  minusIndexFrom(this.dimension, i)
+
+internal inline fun format(value: Double, digits: Int = 4): String {
+    val ten = 10.0
+    val approxOrder = if(value == 0.0) 0 else ceil(log10(abs(value))).toInt()
+    val order = if(
+        ((value % ten) == 0.0) or
+        (value == 1.0) or
+        ((1/value) % ten == 0.0)) approxOrder else approxOrder - 1
+    val lead = value / ten.pow(order)
+    val leadDisplay = round(lead*ten.pow(digits)) / ten.pow(digits)
+    val orderDisplay = if(order == 0) "" else if(order > 0) "E+$order" else "E$order"
+    val valueDisplay = "$leadDisplay$orderDisplay"
+    val res = if(value < 0.0) valueDisplay else " $valueDisplay"
+
+    val fLength = digits + 6
+    val endSpace = " ".repeat(fLength - res.length)
+    return "$res$endSpace"
+}
+
+internal inline fun DoubleTensor.toPrettyString(): String = buildString {
+    var offset = 0
+    val shape = this@toPrettyString.shape
+    val linearStructure = this@toPrettyString.linearStructure
+    val vectorSize = shape.last()
+    val initString = "DoubleTensor(\n"
+    append(initString)
+    var charOffset = 3
+    for (vector in vectorSequence()) {
+        append(" ".repeat(charOffset))
```

Comment:
It's a suboptimal code piece because repeat creates a new `StringBuilder` internally, you can do something like

```
repeat(charOffset) { append(' ') }
```